staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -265,37 +273,26 @@ func (r *crdHandler) removeDeadStorage() {
 		if !found {
 			glog.V(4).Infof("Removing dead CRD storage for %v", s.requestScope.Resource)
 			s.storage.DestroyFunc()
-			delete(storageMap, uid)
```

Comment:
this needed a copy of the map to avoid a race.

Reply:
no need copy, use only one map with lock. Directly delete is ok.

Reply:
I can live with a RWMutex (acutally suggested that before). Would make the code much easier to read. @liggitt I think it was you who proposed this `atomic.Value` construction with "atomic.Value is faster than a RWLock". Can you comment?

Reply:
Quick googling gives https://texlution.com/post/golang-lock-free-values-with-atomic-value/ and https://talks.golang.org/2015/tricks.slide#30. In short: `atomic.Value` is lock free and somewhat faster. But I very much doubt this makes a difference here where we juggling back and forth with json serialization.

  | juggling
-- | --

Reply:
Things that are hardly ever written and read very frequently from lots of threads benefit from atomic value reads

Reply:
I prefer keeping the copy on write approach, and simply modifying this to make a copy and store it

Reply:
What's the advantage from just use one map with RWMutex protect?  From my perspective, it is just do one more copy and atomic op, and does not reduce threads conflicts.

Reply:
> What's the advantage from just use one map with RWMutex protect? From my perspective, it is just do one more copy and atomic op, and does not reduce threads conflicts.

Writes are incredibly rare, and RWMutex can still cause read contention (especially with high numbers of cores)

Reply:
For example, see https://gist.github.com/dim/152e6bf80e1384ea72e17ac717a5000a parallel benchmarks

Reply:
Yes, the result is resonable. Its pure performance about RWMutex and atomic.Value. And they are both `ns` , and not differ too much. In real cases, these locks conflicts are more likely decided by critical sections.

Reply:
In this file, I find the critical sections also contains some logic like `log print`, which is unacceptable in language like c.

Reply:
The point of the critical write sections is that they are exercised very very rarely (only when a CRD changes), so the size of the critical section isn't very important. The reads are exercised extremely frequently, so lockless is benificial there.